ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py

```diff
@@ -0,0 +1,191 @@
+      @ In, clargs, dict, optional, dictionary containing the command-line flags the user can specify in the input (e.g. under the node < Code >< clargstype =0 input0arg =0 i0extension =0 .inp0/ >< /Code >)
+      @ In, fargs, dict, optional, a dictionary containing the auxiliary input file variables the user can specify in the input (e.g. under the node < Code >< fileargstype =0 input0arg =0 aux0extension =0 .aux0/ >< /Code >)
+      @ In, preExec, string, optional, a string the command that needs to be pre-executed before the actual command here defined
+      @ Out, returnCommand, tuple, tuple containing the generated command. returnCommand[0] is the command to run the code (string), returnCommand[1] is the name of the output root
+    """
+    if clargs==None:
+      raise IOError('No input file was specified in clargs!')
+    #check for duplicate extension use
+    usedExts = list(ext[0][0] for ext in clargs['input'].values() if len(ext) != 0)
+    if len(usedExts) != len(set(usedExts)):
+      raise IOError('GenericCodeInterface cannot handle multiple input files with the same extension.  You may need to write your own interface.')
+    for inf in inputFiles:
+      ext = '.' + inf.getExt() if inf.getExt() is not None else ''
+      try:
+        usedExts.remove(ext)
+      except ValueError:
+        pass
+    if len(usedExts) != 0:
+      raise IOError('Input extension',','.join(usedExts),'listed in XML node Code, but not found in the list of Input of <Files>')
+
+    def findSettingIndex(inputFiles,ext):
+      """
+      Find the settings file and return its index in the inputFiles list.
+      @ In, inputFiles, list of InputFile objects
+      @ In, ext, string, extension of the settings.yml file
+      """
+      for index,inputFile in enumerate(inputFiles):
+        if inputFile.getBase() == 'settings' and inputFile.getExt() == ext:
+          return index,inputFile
+      raise IOError('No settings file with extension '+ext+' found!')
+
+    def setOutputDir(settingsFile):
+      """
+      Set the output directory in the settings file.
+      @ In, settingsFile, InputFile object, settings.yml file
+      """
+      # the settings file is the settings.yml file the scenario name is in node
+      # simulation -> scenario_name
+      # the output directory is the directory of the settings file in subdirectory "outputs/$scenario_name"
+      # get the scenario name by reading the settings.yml file
+      scenarioName = None
+      with open(settingsFile.getAbsFile(),'r') as settings:
+        for line in settings:
+          if 'scenario_name' in line:
+            scenarioName = line.split(':')[1].strip()
+            # remove the double quotes
+            scenarioName = scenarioName.replace('"','')
+            break
+      self._outputDirectory = os.path.join(os.path.dirname(settingsFile.getAbsFile()),'outputs',scenarioName)
+      return None
+
+    #prepend
+    todo = ''
```

Comment:
There are some improvements that have been added to https://github.com/idaholab/raven/pull/2151 that need to be merged into this file.